### PR TITLE
ARROW-12594: [C++] Expose the Arrow format version that the library was compiled against

### PR DESCRIPTION
**Pull request by westonpace**
It may be tedious to make sure to update this when the format changes (should probably happen at release time so multiple format changes can be consolidated into one?) but it seems unavoidable.

I considered adding a version enum to `schema.fbs` but I couldn't find any precedent for it and I'm not sure if that is the correct way to go about things anyways.

### CODE REVIEW

File: cpp/CMakeLists.txt [hunk 1]
```diff
@@ -88,10 +88,18 @@ else()
   set(ARROW_FULL_SO_VERSION "${ARROW_SO_VERSION}.${ARROW_VERSION_PATCH}.0")
 endif()
 
+set(ARROW_FORMAT_VERSION "1.0.0")
+set(ARROW_FORMAT_VERSION_MAJOR "1")
+set(ARROW_FORMAT_VERSION_MINOR "0")
+set(ARROW_FORMAT_VERSION_PATCH "0")
```
Comment (cyb70289) [inline cpp/CMakeLists.txt:94]: What about extracting major/minor/patch from version string?
https://github.com/apache/arrow/blob/master/cpp/cmake_modules/FindNumPy.cmake#L83-L87
https://stackoverflow.com/a/18658684/6837980

Comment (westonpace) [reply cpp/CMakeLists.txt]: That works, thanks!

**Comment by nealrichardson** [comment]
Can we add this to R too please 🙏 ?

**Comment by nealrichardson** [comment]
We have [some build info exposed](https://github.com/apache/arrow/blob/master/r/R/arrow-package.R#L137-L170) but maybe there's a more generic source of build info that could be exposed? If so, would you mind just making a JIRA for that?

**Comment by westonpace** [comment]
@nealrichardson ARROW-12602 / #10200

File: cpp/CMakeLists.txt [hunk 2]
```diff
@@ -88,10 +88,20 @@ else()
...
+list(GET _ARROW_FORMAT_VERSION_LIST 1 ARROW_FORMAT_VERSION_MINOR)
+list(GET _ARROW_FORMAT_VERSION_LIST 2 ARROW_FORMAT_VERSION_PATCH)
+
 message(STATUS "Arrow version: "
                "${ARROW_VERSION_MAJOR}.${ARROW_VERSION_MINOR}.${ARROW_VERSION_PATCH} "
                "(full: '${ARROW_VERSION}')")
 message(STATUS "Arrow SO version: ${ARROW_SO_VERSION} (full: ${ARROW_FULL_SO_VERSION})")
+message(STATUS "Arrow Format version: ${ARROW_FORMAT_VERSION_MAJOR}."
+               "${ARROW_FORMAT_VERSION_MINOR}.${ARROW_FORMAT_VERSION_PATCH} "
+               "(full: '${ARROW_FORMAT_VERSION}')")
```
Comment (kou) [inline cpp/CMakeLists.txt:104]: We can just show `${ARROW_FORMAT_VERSION}` here. Because `${ARROW_FORMAT_VERSION}` never have `-SNAPSHOT`.

**Comment by westonpace** [comment]
Per discussion on the ML I am going to abandon this PR.  Interoperability should be considered on a per-feature basis and not a per-version basis as implementations are not required to fully implement any one version of the spec.   Publicly exposing this version could only lead to confusion.

**Pull request closed without merging**